### ISSUE TITLE
3.8.0 Release Plan

### ISSUE THREAD

**Issue by x87** [sannybuilder/dev#130]
Download: [v3.8.0](https://github.com/sannybuilder/dev/releases/tag/v3.8.0)
--

## Features
- [x] theme customization
  - [x] #3 
  - [x] [theme editor](https://docs.sannybuilder.com/options/theme-editor)
  - [x] convert existing coloring configuration to the custom (user) theme
  - [x] autocomplete list coloring
  - [x] [community themes](https://github.com/sannybuilder/themes)
  - [x] [update language files](https://github.com/sannybuilder/translations/pull/13)
  - [x] documentation
- [x] update translations
  - [x] [Chinese](https://github.com/sannybuilder/translations/pull/10)
  - [x] [Hungarian](https://github.com/sannybuilder/translations/pull/12)
  - [x] [Spanish](https://github.com/sannybuilder/dev/issues/3#issuecomment-854329662)
 - [x] integrate with https://library.sannybuilder.com/
   - [x] https://github.com/sannybuilder/core/pull/8
   - [x] automatic/on-demand update
   - [x] clicking the bottom panel navigates to the library
   - [x] documentation
 - [x] #33 
 - [x] #108 
 - [x] #93
   - [x] [consistent option names](https://github.com/sannybuilder/dev/issues/93#issuecomment-864450701)
 - [x] new debug option "SKIP_EXTRA_INFO" to ignore custom information attached to the compiled script
   - [x] [documentation](https://docs.sannybuilder.com/console#skip_extra_info)
- [x] #140 
 - [x] #154 
 - [x] [update CLEO+ plugin](https://github.com/JuniorDjjr/CLEOPlus/releases)
   - [x] Sanny Builder installs CLEO+ plugin along with other CLEO plugins
 - [x] the installer action "Associate with Sanny Builder" changes the Windows Registry for the current user only (https://github.com/sannybuilder/dev/issues/43#issuecomment-670219723)
 - [x] update hotkeys
   - [x] #26
   - [x] #122
   - [x] [make "Run" menu hotkeys customizable](https://github.com/sannybuilder/dev/issues/26#issuecomment-891425137)

## Bug Fixes
- LCS/VCS support
  - [x] #85 
  - [x] #164
  - [x] #165
  - [x] compiler adds 8 redundant bytes in a headless script in LCS/VCS modes
  - [x] incorrect data type picked for labels in LCS/VCS modes
- [x] #117
- [x] #131 
- [x] #139 
- [x] #146
- [x] #149 
- [x] disassembler incorrectly treats label type in classes
- [x] compiler fails on a deprecated "extended" param (reported [here](https://sannybuilder.com/forums/viewtopic.php?id=6109))
- [x] [incorrect status of the language service in the status bar](https://github.com/sannybuilder/dev/issues/130#issuecomment-852187718)
- [x]  "New CLEO script" menu item only accessible when another document is open 
- [x] [template list displays duplicate entries](https://github.com/sannybuilder/dev/issues/148#issuecomment-864458204)
- [x] CLEO installation from the installer does not work

**Comment by x87** [comment]
3.8.0-alpha.0 is available for preview and testing
https://github.com/sannybuilder/dev/releases/tag/v3.8.0-alpha.0

Please report any issues found in this version.

**Comment by wmysterio** [comment]
>3.8.0-alpha.0 is available for preview and testing

My test script compiled without problems, the game didn't crash. I checked out some of the new options. So far I have not seen anything suspicious.

I have developed two [new themes](https://drive.google.com/file/d/1hOdkIx0X5zg-A6cUkOs5AnmAOO6EcCJg/view) to check out (ini).

**Comment by x87** [comment]
>Templates are not added to my code (CTRL+J). When I press Enter or click, the template name is inserted, not all of its content.

thank you for checking 👍

**Comment by VitalRus95** [comment]
All in all, the update is great! I liked the new features, especially theme editor and opcode hints. The existing colour scheme settings have transferred to the new version nice, the only exception being the editor background, I had to set it again but that's a minor inconvenience. That's how SB looks with 3.8.0 alpha:
![изображение](https://user-images.githubusercontent.com/73489604/120337100-affb6480-c2fb-11eb-91e4-6b5baad5ce9d.png)

I also tested compiling a script from command line, works fine now even without `{$NOSOURCE}`, thank you, @x87!

As for small bugs, there are some:
- just like @wmysterio, I can't use templates, only their names are being inserted into the code;
- I have several tabs with different GTA games, and when, just having started SB, I switched from VC script tab to SA script tab, the status of language service remains “Scanning” until I manually choose the corresponding game again, even though it is already chosen in the bottom right corner and the script compiles without errors;
- again tabs issues: if you have several tabs, then close the editor, then open a script or source from Windows Explorer, the tabs will all be replaced with the script you opened, which is a bit annoying;
- when hovering over bottom panel, the hints like “Double click to open SB library in your browser” flicker. I also find the font to be quite small there, maybe it would be cool to have pop-up hints of the regular text size (which can be set by the user) right above/below the selected opcode when hovering your mouse over it. Yet it's not a big problem and I don't know if this idea can be implemented.

Other than that, the new version is a big step forward, I like it! By the way, will there be an offline version of the opcode library just in case? Good luck in further development and thanks for making SB better!

**Comment by x87** [comment]
@VitalRus95 thank you for your support! I will look into the issues you've reported.

> will there be an offline version of the opcode library just in case
in the current implementation the opcode library is completely offline. There are gta3.json/vc.json/sa.json files in the respective mode directories (controlled by the new `<library>` property in modes.xml). I plan to make an automatic and on-demand update of those files from within the IDE. Also you can download newer versions from https://github.com/sannybuilder/library and update your local installation.

https://github.com/sannybuilder/library/blob/master/gta3/gta3.json
https://github.com/sannybuilder/library/blob/master/vc/vc.json
https://github.com/sannybuilder/library/blob/master/sa/sa.json

**Comment by x87** [comment]
Published 3.8.0-alpha.1
https://github.com/sannybuilder/dev/releases/tag/v3.8.0-alpha.1

**Comment by wmysterio** [comment]
> Published 3.8.0-alpha.1

I have not noticed any problems. A later I will test the CLI options.

**Comment by VitalRus95** [comment]
Templates now work perfectly fine, thanks! I haven't tested opening a script or source from Windows Explorer, but the rest of the issues I listed above are still present. Also, there's a little problem I encountered as early as in 3.7.0 version of SB:
![изображение](https://user-images.githubusercontent.com/73489604/121198115-8c479980-c87a-11eb-8285-a2899c99bce9.png)

However, when I click OK, thus closing the error windows, SB continues working without any errors. Strange. But at least there's nothing that breaks the coding, just some inconvenience. I'm glad the programme is getting better and receives new nice features!

**Comment by x87** [comment]
>Also, there's a little problem I encountered as early as in 3.7.0 version of SB
is this reproducible in 3.8.0? Can you share the sanny.log ?

**Comment by VitalRus95** [comment]
Yes, it keeps happening in 3.8.0. I would willingly share the .log, but I don't have one near sanny.exe, unfortunately. All I have is these three errors at the same time:
![изображение](https://user-images.githubusercontent.com/73489604/121264868-870d3d80-c8c0-11eb-8c0f-aae258d387e9.png)
When I close them, SB loads and works like nothing happened.

**Comment by x87** [comment]
Published 3.8.0-alpha.2
https://github.com/sannybuilder/dev/releases/tag/v3.8.0-alpha.2

**Comment by VitalRus95** [comment]
@x87, thanks for the new alpha!

Two problems have been **fixes**:
- startup error messages;
- “Scanning” status of the language service staying on screen forever.

Two little things **remain**:
- pop-up hints of the bottom panel still flicker;
- tabs get cleared and replaced by a file you open from Windows Explorer.

By the way, I noticed that templates are reset to standard ones after installing new versions, which is also an inconvenience, since I, for one, have my own CLEO template and slightly modified standard templates of conditions and loops. It would be cool to have them saved until a serious change is needed.

A thought crossed my mind about templates: maybe if a user creates their own CLEO template, it could be used by SB in “Create new CLEO script” menu option. But these are just suggestions, not a high priority thing to consider.

I like how fast you react to the feedback, good luck with further development! I'll also try to spread a word about the new features that will appear in 3.8.0 to make people more interested in updating. :)

**Comment by x87** [comment]
>A thought crossed my mind about templates: maybe if a user creates their own CLEO template, it could be used by SB in “Create new CLEO script” menu option.

this is exactly what should happen. New CLEO script menu is a combo of two actions: Open a new document and Apply a template with the name "cleo".

**Comment by VitalRus95** [comment]
Hmm, it should but in my case it doesn't for some reason. I have the following template:
`cleo=
={$CLEO} // Script by Vital (Vitaly Ulyanov)
=script_name '|'
=
=const
= plr = $PLAYER_ACTOR
= pgr = $PLAYER_GROUP
= mis = $ONMISSION
=end // const
=
=while true
=    
=end // while
=
=terminate_this_custom_script`

But whenever I click “New CLEO script” in SB, I only get
`{$CLEO .cs}
0000:



0A93: terminate_this_custom_script`

**Comment by x87** [comment]
>I have the following template:

Which file is that? Templates are of two kinds: shared and exclusive. https://docs.sannybuilder.com/edit-modes/code-templates#types-of-templates

You might have added your custom template to `data\templates.txt` and it has a lower priority than the template with the same name `cleo` defined in `data\sa\templates.txt`

**Comment by VitalRus95** [comment]
I intended the template to be shared, as it is suitable for all GTA games I write scripts for. But yes, you are right, moving the same code to `sa\templates.txt` helped, now in SA mode it inserts my template. So, I think now I'll have to copy it to each game's `templates.txt` to make it work for GTA 3, VC, and SA at once. Thanks for the prompt!

**Comment by x87** [comment]
@VitalRus95 also you can delete all exclusive versions of "cleo" template and keep only the shared one if that works for you.

**Comment by x87** [comment]
I haven't found a good way to deal with user modifications yet. I know that it could be painful to migrate from one version to another if you heavily modified data files. I can open a new discussion for this topic, so we evaluate what options do we have.

**Comment by x87** [comment]
Published 3.8.0-alpha.3
https://github.com/sannybuilder/dev/releases/tag/v3.8.0-alpha.3

**Comment by VitalRus95** [comment]
I have a small suggestion, don't know if that's possible to implement, but it would be nice. I noticed that you can't call class commands from a constant. For instance, if you declare `3@` as an `Actor` and assign a constant `swt` to the value `3@`, then you can use
```
3@.Create(…)
```
yet you won't be able use
```
swt.Create(…)
```
The latter, I think, may make constants even more helpful in SB (again, if possible and reasonable).

**Comment by x87** [comment]
Published 3.8.0-alpha.4
https://github.com/sannybuilder/dev/releases/tag/v3.8.0-alpha.4

**Comment by VitalRus95** [comment]
So far 3.8.4 is good, yet I found two issues:
- I created a new script (thus adding a new tab) and after restarting the programme this tab was gone; I reopen the script from the “File” menu, but its tab disappears again after restarting SB;
- last opened documents section of the “File” menu doesn't work and shows nothing.

**Comment by x87** [comment]
@VitalRus95 I can't reproduce it on my machine with 3.8.0-alpha.4

can you share your settings.ini

**Comment by VitalRus95** [comment]
@x87, sure, here's my settings.ini.

As for the file's path, it only has English letters.
[settings.txt](https://github.com/sannybuilder/dev/files/6696756/settings.txt)

**Comment by x87** [comment]
@VitalRus95 as I mentioned in https://github.com/sannybuilder/dev/issues/146#issuecomment-857282090 it appears thst your settings.ini does not have newer configuration entries. Can you investigate, why it does not have a theme name and renamed options from #93 ? Do you have multiple sanny installations? Sid you install sanny in a UAC protected directory such ax Program Files? Is this settings.ini read-only?

**Comment by VitalRus95** [comment]
Hm, as for the theme, the file contains the line `Editor::Theme=Dark (by Vital)`. And is not marked as read-only.

My `settings.ini` is located in `C:\Users\ulian\AppData\Local\VirtualStore\Program Files (x86)\Sanny Builder 3\data`.
Sanny Builder's folder is `C:\Program Files (x86)\Sanny Builder 3`.

I have one SB installation, yet I usually do not uninstall previous versions before installing new ones. I did do a complete reinstallation when I noticed the issue, but it didn't help.

Does it all mean that it's buggy only because of `Program Files (x86)`? Why on earth does the OS prompts to install programmes there if they keep breaking this way…

**Comment by x87** [comment]
@VitalRus95 check #43 for details

**Comment by VitalRus95** [comment]
@x87, thanks for the link! I set SB shortcut to run as admin, and it seems to have resolved the issue! Also, now I found `settings.ini` in SB's `data` folder. Hope running as admin would prevent further conflicts with Windows. Never thought the system is that quirky.

**Comment by x87** [comment]
You can install SB anywhere outside of `Program Files (x86)` to avoid the need in admin mode.

**Comment by XMDS** [comment]
@x87  3.8.0-alpha.4 When I choose a translation language, after closing SB, opening SB again will restore to the default English language

**Comment by x87** [comment]
> @x87 3.8.0-alpha.4 When I choose a translation language, after closing SB, opening SB again will restore to the default English language

Do you run Sanny as admin? See discussion above

**Comment by x87** [comment]
> 
> 
> @x87 3.8.0-alpha.4 When I choose a translation language, after closing SB, opening SB again will restore to the default English language

there is an issue in alpha.4 with renamed options. When settings.ini contains both the old and the new name, the former takes a priority. A workaround is to delete your settings.ini and re-run Sanny.

**Comment by x87** [comment]
Published v3.8.0-beta.0

https://github.com/sannybuilder/dev/releases/tag/v3.8.0-beta.0

**Comment by VitalRus95** [comment]
It seems creating a new CLEO script with a custom template works fine, now it's more intuitive, thank you very much for this! But I noticed this version doesn't seem to recognise neither tabs nor several (in my case four) spaces, substituting them with just one space. Very strange. In templates (both custom and predefined) there are four spaces in some templates, they don't work any more.

![изображение](https://user-images.githubusercontent.com/73489604/126882306-61d61db1-a9cc-4165-9066-e00d2607a2b0.png)

**Comment by x87** [comment]
>this version doesn't seem to recognise neither tabs nor several (in my case four) spaces, substituting them with just one space.

thanks for the report, confirmed.

**Comment by x87** [comment]
Published v3.8.0-beta.1

https://github.com/sannybuilder/dev/releases/tag/v3.8.0-beta.1

**Comment by wmysterio** [comment]
> Published v3.8.0-beta.1

Perhaps this is the way it should be, but I noticed that the path to the game is not saved after installation and I have to specify it again in the program settings (the "CLEO" button is highlighted in red, although I also installed it and indicated the path to the game).

**Comment by x87** [comment]
>the path to the game is not saved after installation

can you confirm if there is an exe file inside the directory that you chose? (gta3.exe, gta-vc.exe or gta[_-]sa.exe)

>the "CLEO" button is highlighted in red, although I also installed it and indicated the path to the game

you installed it separately from Sanny Builder? Is there cleo.asi file in the game directory?

**Comment by wmysterio** [comment]
> can you confirm if there is an exe file inside the directory that you chose? (gta3.exe, gta-vc.exe or gta[_-]sa.exe)

I renamed the specified file ("gta_sa.exe"), since the game starts with this name only once.

>you installed it separately from Sanny Builder? Is there cleo.asi file in the game directory?

I installed CLEO only with the Sunny Builder installer. The file "cleo.asi" existed. Shouldn't it be overwritten?

**Comment by x87** [comment]
sorry, I'm a bit confused about the order of the steps you made. I see multiple issues here.

Installing CLEO during Sanny Builder installation does not work. Even if you specify the correct directory and select the option install CLEO it does nothing, this code is broken. So after Sanny Builder installation there should be no changes made to the game directory.

I have fixed this and now Sanny Builder's installer should properly copy CLEO files to the game directory.


Selecting game directory during install should work and Sanny Builder should keep this path in the settings. Can you make some testing on your end and confirm that it does not happen to you? Try to make a fresh install, select the game directory and then launch Sanny Builder and see if it has the path in the options?

**Comment by wmysterio** [comment]
@x87 I did tests on "3.8.0-beta.1" and they showed that the path to the game is not saved if the folder does not contain "gta_sa.exe". In all other cases, the path disappears from the settings. If this behavior is expected, then there will be no more questions. I just wondered why the "gta_sa.exe" plays a strong role when only the directory is needed.

**Comment by x87** [comment]
@wmysterio this is because when you select a directory Sanny Builder should somehow identify the game. It can be GTA III, VC or SA. Exe file is a good marker.

Possibly this option can be rethough, and instead of asking for one directory the installer could offer 3 different fields, one for each game. Then there won't be a need for an executable in the directory.

**Comment by wmysterio** [comment]
@x87 Hmm, now I installed SB without CLEO and with "gta_sa.exe". Result: edit mode "gta sa 1.0", but the path to the game was not saved o_0. Some strange behavior.

**Comment by x87** [comment]
@wmysterio can you share settings.ini from this installation?

**Comment by wmysterio** [comment]
@x87 sure, [settings.ini](http://ru-script.3dn.ru/DELETE_LATER/settings.ini).

**Comment by x87** [comment]
@wmysterio I don't see any game related settings. I looks like the installer was unable to identify the game for that directory. I tried to reproduce the same: created a new folder and added file gta_sa.exe in it. Then I run 3.8.0-beta.1 installer, selected that folder, skipped CLEO library installation and did not run Sanny Builder after installation. My settings ini has the following lines:

```
[SA]
GamePath=C:\gta-sa
[Main]
EditMode=2
[Lang]
Lang=1033

```

Can you do one more experiment. Repeat all steps again, but don't launch Sanny Builder automatically after installation. Go to its folder and share what's in the settings.ini. There should be only a few lines.

> Result: edit mode "gta sa 1.0"

this is a default choice if no other edit mode was set.

**Comment by wmysterio** [comment]
@x87 File after these steps, but without autorun:

```text
[SA]
GamePath=D:\Programm\GTA_SA
[Main]
EditMode=2
[Lang]
Lang=1049
```

Hmm, tried another identical test, but with autostart and "settings.ini" it became identical to the lines above o_0. Very strange. I wonder if other users have similar incidents or is it just me? ..

**Comment by x87** [comment]
@wmysterio so, first of all, there is another issue that I found looking into our initial settings.ini. Installer generates the file with `EditMode` value set to an integer, whereas Sanny Builder expects this option to be a string: https://docs.sannybuilder.com/cli#mode The correct values are 'gta3', 'vc', 'sa'. I have fixed this.

As for your observations, I can't say for sure what is going on, but maybe you have multiple Sanny Builder installations and you are installing SB in one directory, but checking settings.ini in the other directory?

**Comment by wmysterio** [comment]
> As for your observations, I can't say for sure what is going on, but maybe you have multiple Sanny Builder installations and you are installing SB in one directory, but checking settings.ini in the other directory?

Before installing, I always delete the "Sanny Builder 3" folder and I will install all new versions from scratch. I also have a "Sanny Builder 3.5.2" folder, but she's unlikely to be involved in this. I'll do some more tests for now. I don’t know what’s going on yet.

**Comment by wmysterio** [comment]
Rapported by @VitalRus95 . The error is also reproducible for me.:

> again tabs issues: if you have several tabs, then close the editor, then open a script or source from Windows Explorer, the tabs will all be replaced with the script you opened, which is a bit annoying;

The behavior of the tabs remains the same.

**Comment by x87** [comment]
@wmysterio we have a ticket #151
